File: core/src/main/java/hunternif/voxarch/storage/BlockData.java
package hunternif.voxarch.storage;

import hunternif.voxarch.util.Direction;

/**
 * Contains block id, metadata and {@link Direction}. The orientation
 * should modify the metadata in overridden method {@link #setOrientation}.
 * @author Hunternif
 */
public class BlockData {
	private int id;
	private int metadata;
	
	private Direction orient = null;
	
	public BlockData(int id) {
		this(id, 0);
	}
	
	public BlockData(int id, int metadata) {
		this.id = id;
		this.metadata = metadata;
	}
	
	public void setId(int id) {
		this.id = id;
	}
	public int getId() {
		return id;
	}
	
	public void setMetadata(int metadata) {
		this.metadata = metadata;
	}
	public int getMetadata() {
		return metadata;
	}
	
	public boolean hasOrientation() {
		return orient != null;
	}
	
	/** Rotate the Direction (if not NONE) counterclockwise by the
	 * specified angle. */
	public void rotate(double angle) {
		if (!hasOrientation()) return;
		setOrientation(orient.rotate(angle));
	}

	public Direction getOrientation() {
		return orient;
	}
	
	/** Override this method to set metadata appropriately. */
	public void setOrientation(Direction orient) {
		this.orient = orient;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof BlockData)) return false;
		if (obj == this) return true;
		return ((BlockData)obj).id == id && ((BlockData)obj).metadata == metadata;
	}
	
	@Override
	public BlockData clone() {
		return new BlockData(id, metadata);
	}
	
	@Override
	public String toString() {
		return "id: " + id + " " + orient.name() + " meta: " + metadata;
	}
}

File: core/src/main/kotlin/hunternif/voxarch/storage/BlockData.kt
package hunternif.voxarch.storage

import hunternif.voxarch.util.Direction

/**
 * Contains block [key] and [orientation].
 * (I tried to make this an interface, but the generics are too annoying.)
 */
open class BlockData(
    val key: String,
    val orientation: Direction? = null
) : IVoxel {

    // Legacy constructor
    constructor(key: String) : this(key, null)

    /**
     * Returns an instance of this block that's rotated and facing the Direction
     * (if not NONE) counterclockwise by the specified angle.
     */
    fun rotate(angle: Double): BlockData {
        return if (angle == 0.0 || orientation == null) this
        else orient(orientation.rotate(angle))
    }

    /** Returns an instance with the same [key] and given [newOrientation] */
    fun orient(newOrientation: Direction): BlockData {
        return if (orientation == newOrientation) this
        else getInstance(newOrientation)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is BlockData) return false
        return if (other === this) true
        else other.key == key && other.orientation == orientation
    }

    /** Returns an instance with the same [key] and given orientation. */
    protected open fun getInstance(orientation: Direction): BlockData {
        return BlockData(key, orientation)
    }

    override fun toString(): String {
        return "id: " + key + " " + orientation?.name
    }

    override fun hashCode(): Int {
        var result = key.hashCode()
        result = 31 * result + (orientation?.hashCode() ?: 0)
        return result
    }
}